sdk/lib/opentelemetry/sdk/trace/samplers/probability_sampler.rb

```diff
@@ -8,86 +8,70 @@ module OpenTelemetry
   module SDK
     module Trace
       module Samplers
-        # A {OpenTelemetry::Trace::Samplers::Sampler} where the probability of
-        # sampling a trace is equal to that of the specified probability.
-        class ProbabilitySampler < OpenTelemetry::Trace::Samplers::Sampler
-          class << self
-            private :new # rubocop:disable Style/AccessModifierDeclarations
+        # @api private
+        #
+        # Implements sampling based on a probability.
+        class ProbabilitySampler
+          def initialize(probability, hints:, ignore_parent:, apply_to_root_spans:, apply_to_remote_parent:, apply_to_all_spans:)
+            @probability = probability
+            @id_upper_bound = format('%016x', (probability * (2**64 - 1)).ceil)
+            @result_from_hint = hints.map { |hint| [hint, Result.new(decision: hint)] }.to_h.freeze
+            @ignore_parent = ignore_parent
+            @apply_to_root_spans = apply_to_root_spans
+            @apply_to_remote_parent = apply_to_remote_parent
+            @apply_to_all_spans = apply_to_all_spans
+          end
 
-            # Returns a new {OpenTelemetry::Trace::Samplers::Sampler}. The
-            # probability of sampling a trace is equal to that of the specified
-            # probability.
-            #
-            # @param [Numeric] probability The desired probability of sampling.
-            #   Must be within [0.0, 1.0].
-            # @raise [ArgumentError] if probability is out of range
-            # @return [OpenTelemetry::Trace::Samplers::Sampler]
-            def create(probability)
-              raise ArgumentError, 'probability must be in range [0.0, 1.0]' unless (0.0..1.0).include?(probability)
+          # @api private
+          #
+          # Callable interface for probability sampler. See {Samplers}.
+          def call(trace_id:, span_id:, parent_context:, hint:, links:, name:, kind:, attributes:)
+            take_hint(hint) ||
```

Comment:
This is pretty gnarly logic, but nicely organized into understandable, well-named methods.